src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,218 @@
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal static class PercentEncodingHelpers
+    {
+        private static readonly char[] HexChars = new[]
+        {
+            '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
+        };
+        internal const int MaxUnicodeCharsReallocate = 40; // Maximum batch size when working with unicode characters
+        private const int MaxUtf8BytesPerUnicodeChar = 4;
+        private const int AsciiMaxValue = 127;
+
+        // From https://github.com/grpc/grpc/blob/324189c9dc540f0693d79f02dcb8c5f9261b535e/src/core/lib/slice/percent_encoding.cc#L31
+        private static readonly byte[] PercentEncodingUnreservedBitField =
+        {
+            0x00, 0x00, 0x00, 0x00, 0xdf, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xff, 0xff, 0xff, 0x7f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+        };
+
+        public static string PercentEncode(string value)
+        {
+            // Count the number of bytes needed to output this string
+            var encodedLength = 0;
+            for (var i = 0; i < value.Length; i++)
+            {
+                var c = value[i];
+                if (c > AsciiMaxValue)
+                {
+                    // Get additional unicode characters
+                    var unicodeCharCount = GetUnicodeCharCount(value, i);
+
+                    var unicodeCharacters = Encoding.UTF8.GetByteCount(value.AsSpan(i, unicodeCharCount));
+                    encodedLength += unicodeCharacters * 3;
+                    i += unicodeCharCount - 1;
+                }
+                else
+                {
+                    bool unres = IsUnreservedCharacter(c);
+                    encodedLength += unres ? 1 : 3;
+                }
+            }
+
+            // Return the string if no encoding is required
+            if (value.Length == encodedLength)
+            {
+                return value;
+            }
+
+            // Encode
+            return string.Create(encodedLength, value, Encode);
+
+            static void Encode(Span<char> span, string s)
+            {
+                Span<byte> unicodeBytesBuffer = stackalloc byte[MaxUnicodeCharsReallocate * MaxUtf8BytesPerUnicodeChar];
+
+                var writePosition = 0;
+                for (var i = 0; i < s.Length; i++)
+                {
+                    var current = s[i];
+                    if (current > AsciiMaxValue)
+                    {
+                        // Get additional unicode characters
+                        var unicodeCharCount = GetUnicodeCharCount(s, i);
+
+                        var numberOfBytes = Encoding.UTF8.GetBytes(s.AsSpan(i, unicodeCharCount), unicodeBytesBuffer);
```

Comment:
The behavior of this method is that invalid UTF-16 data (such as unpaired `'\uD800'` chars) will get converted to the UTF-8 byte sequence `[ EF BF BD ]`, then escaped as `"%EF%BF%BD"`. The receiver will decode it as the replacement char � (`'\uFFFD'`).

This is _probably_ appropriate behavior for this scenario, but I wanted to call it out explicitly just so you could confirm that was your intention.

Reply:
This is intended. There are unit tests below that explicitly test for this - https://github.com/grpc/grpc-dotnet/pull/223/files#diff-4269317ee063a7595ca58194903e29f6R70

I'll add a comment here that discusses this possibility.